### PR TITLE
Add small sr test suite

### PR DESCRIPTION
**Pull request by Meggison**
## 📝 Description

**⚠️ Issue Link Required**: This PR must be linked to an approved and assigned issue. See [Contributing Guide](https://github.com/kornia/kornia/blob/main/CONTRIBUTING.md#pull-request) for details.

**Fixes/Relates to:** (#3553)

**Important**:
- Ensure you are assigned to the linked issue before submitting this PR
- This PR should strictly implement what the linked issue describes
- Do not include changes beyond the scope of the linked issue

---

### Test Suite Structure

**`TestSmallSRNet` (44 tests - core model):**
- [x] Smoke tests for upscale factors 2x, 3x, 4x with multiple batch sizes
- [x] Exception tests for invalid inputs and edge cases
- [x] Cardinality tests verifying output shapes for various input dimensions
- [x] Pretrained weight loading verification (upscale_factor=3)
- [x] Forward pass validation ensuring non-zero, finite outputs
- [x] Gradient checks using `gradcheck` for backpropagation correctness
- [x] Torch.compile compatibility tests with `torch_optimizer` fixture

**`TestSmallSRNetWrapper` (46 tests - RGB wrapper):**
- [x] Smoke, exception, and cardinality tests mirroring SmallSRNet coverage
- [x] Color space conversion validation (RGB→YCbCr→RGB pipeline)
- [x] Batch consistency tests comparing batch vs. individual processing
- [x] RGB processing end-to-end validation
- [x] Gradient checks and torch.compile compatibility

**Testing Standards:**
- [x] Follows `BaseTester` pattern (consistent with KimiVL, Siglip2, SAM3)
- [x] Uses parametrized fixtures for device/dtype combinations
- [x] Comprehensive edge case coverage

---

### Unit Tests
- [x] **90 test cases** covering both `SmallSRNet` and `SmallSRNetWrapper`
- [x] Tests include: smoke, exception, cardinality, pretrained loading, gradcheck, dynamo
- [x] All tests follow repository `BaseTester` pattern
- [x] Device/dtype parametrization (cpu, cuda, float32, float64)

### Manual Verification
```bash
# With bug fix applied from dependency PR:
pixi run test tests/models/test_small_sr.py -v
# Result: ✅ 90/90 tests passing

pixi run lint
# Result: ✅ All pre-commit hooks passing

pixi run typecheck  
# Result: ✅ Type checking passed
```

### Performance/Edge Cases
- [x] **Different upscale factors**: Tests cover 2x, 3x, 4x super-resolution
- [x] **Various input sizes**: 64×64, 128×128, 224×224 resolutions
- [x] **Batch sizes**: Tests with batch sizes 1, 2, 4
- [x] **Color space handling**: RGB↔YCbCr conversion accuracy
- [x] **Pretrained weights**: Verifies weight loading and model instantiation
- [x] **Invalid inputs**: Tests exception handling for wrong input channels
- [x] **Gradient computation**: Validates backpropagation through all layers
- [x] **Compiler compatibility**: Ensures torch.compile works correctly

---

---

## 🧪 How Was This Tested?
- [ ] **Unit Tests:** (List new/updated tests)
- [ ] **Manual Verification:** (Describe the steps you took)
- [ ] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations)
- [x] My code follows the existing style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with the bug fix applied
- [x] Pre-commit hooks pass (ruff, codespell, license headers)
- [x] Tests follow `BaseTester` pattern used in repository

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.

**Workflow recommendation:**
1. Merge the bug fix PR first (#3550)
2. Then this comprehensive test suite can be validated and merged

### CODE REVIEW

**Comment by edgarriba** [comment]
First-pass review ✅

The test suite looks solid — good coverage across shapes, dtypes, devices, and wrapper logic. A couple of notes:

- The `weight_init` refactor in `small_sr.py` is a nice improvement (using `isinstance` rather than hard-coded attribute access). The out_channels heuristic to detect conv4 is a bit fragile — it works for the current architecture but will silently break if the model changes. A small comment explaining the logic would help future readers.
- The tutorials CI failure is a known infra issue (#3566) and is unrelated to this PR — do not let it block the review.

LGTM from my side — @sidd-27 or @cjpurackal can you give final approval and merge?

**Comment by sidd-27** [comment]
@Meggison  could you rebase from main?
